grabbit/src/test/groovy/com/twcable/grabbit/server/batch/steps/jcrnodes/JcrNodesProcessorSpec.groovy

```diff
@@ -69,4 +75,136 @@ class JcrNodesProcessorSpec extends Specification {
         propertyString.type == STRING
     }
 
+    def "Can marshall a JCR Node with mandatory child node to a Protobuf Message"() {
+        given:
+        // Setup parent node as nt:file which has mandatory child node
+        JcrNode hierarchyNode = Mock()
+        hierarchyNode.path >> "default.groovy"
+        JcrProperty primaryProperty = Mock()
+        primaryProperty.type >> 1
+        primaryProperty.name >> JcrConstants.JCR_PRIMARYTYPE
+        primaryProperty.value >> new ValueImpl(JcrConstants.NT_FILE)
+        PropertyIterator propertyIterator = new PropertyIteratorImpl([primaryProperty])
+        hierarchyNode.properties >> propertyIterator
+
+        NodeDefinition nodeDefinition = Mock()
+        hierarchyNode.definition >> nodeDefinition
+
+        NodeType nodeType = Mock()
+        NodeDefinition childNodeDefinition = Mock()
+        childNodeDefinition.mandatory >> true
+        nodeType.childNodeDefinitions >> [childNodeDefinition]
+        hierarchyNode.primaryNodeType >> nodeType
+
+        // Setup child node "default.groovy/jcr:content", which is mandatory child node and a property type as nt:resource
+        JcrNode childMandatoryNode = Mock()
+        childMandatoryNode.path >> "default.groovy/jcr:content"
+        JcrProperty childPrimaryProperty = Mock()
+        childPrimaryProperty.type >> 1
+        childPrimaryProperty.name >> JcrConstants.JCR_PRIMARYTYPE
+        childPrimaryProperty.value >> new ValueImpl(JcrConstants.NT_RESOURCE)
+        PropertyIterator childPropertyIterator = new PropertyIteratorImpl([childPrimaryProperty])
+        childMandatoryNode.properties >> childPropertyIterator
+        childMandatoryNode.definition >> childNodeDefinition
+
+        childMandatoryNode.parent >> hierarchyNode
+
+        NodeType childNodeType = Mock()
+        childNodeType.childNodeDefinitions >> [ Mock(NodeDefinition)]
+        childMandatoryNode.primaryNodeType >> childNodeType
+        NodeIterator nodeIterator = new NodeIteratorImpl([childMandatoryNode]);
+        hierarchyNode.getNodes() >> nodeIterator
+
+        when:
+        JcrNodesProcessor jcrNodesProcessor = new JcrNodesProcessor()
+        NodeProtos.Node nodeProto = jcrNodesProcessor.process(hierarchyNode)
+
+        then:
+        nodeProto != null
+        nodeProto.name == "default.groovy"
+        // Verified parent node as nt:file which has mandatory child node
+        nodeProto.properties.propertyList.any { it.name == JcrConstants.JCR_PRIMARYTYPE && it.value.stringValue == JcrConstants.NT_FILE}
+        // Verified child node "default.groovy/jcr:content", which is mandatory child node and a property type as nt:resource
+        nodeProto.mandatoryChildNodeList.any { it.name == "default.groovy/jcr:content" && it.properties.propertyList.any { it.name == JcrConstants.JCR_PRIMARYTYPE && it.value.stringValue == JcrConstants.NT_RESOURCE}}
+    }
+
+    // Need to implement the NodeIterator as neither the Mock nor the Jackalope was helpful
```

Comment:
I don't think you need to implement your own Node & Property iterators ... you can leverage the Jackrabbit Commons' `NodeIteratorAdapter` and `PropertyIteratorAdapter` here (that can convert normal iterators to `NodeIterator` or `PropertyIterator`)

Example : in our case : 

```groovy
hierarchyNode.getNodes() >> new NodeIteratorAdapter([childMandatoryNode].iterator())
//...
childMandatoryNode.properties >> new PropertyIteratorAdapter([childPrimaryProperty].iterator())
//...
//etc
```